BaristaSwift/Common/Common/Base/Store.swift

```diff
@@ -89,95 +88,87 @@ public class Store<objectType: StoreProtocol> {
         }
         return results
     }
-
-    public func upsertEntries(jsonResponse: Any, completion: SyncCompletion = nil) {
-        let dataRows = (jsonResponse as! NSDictionary)["records"] as! [NSDictionary]
-        SalesforceSwiftLogger.log(type(of:self), level:.debug, message:"request:didLoadResponse: #records: \(dataRows.count)")
-        store.upsertEntries(dataRows, toSoup: objectType.objectName)
-        completion?(nil)
-    }
-    
-    public func upsertEntries<T:StoreProtocol>(record: T, completion: SyncCompletion = nil) {
-        store.upsertEntries([record.data], toSoup: T.objectName)
-        completion?(nil)
-    }
     
-    public func upsertNewEntries<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
+    public func locallyCreateEntry(entry: objectType) -> objectType {
+        var record: objectType = entry
         record.local = true
         record.locallyCreated = true
-        record.objectType = T.objectName
-        store.upsertEntries([record.data], toSoup: T.objectName)
-        completion?(nil)
+        record.objectType = objectType.objectName
+        return objectType.from(store.upsertEntries([record.data], toSoup: objectType.objectName))
     }
     
-    public func deleteEntry<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
-        record.locallyDeleted = true
-        syncEntry(entry: record, completion: completion)
+    public func locallyUpdateEntry(entry: objectType) -> objectType {
+        var record: objectType = entry
+        record.local = true
+        record.locallyUpdated = true
+        return objectType.from(store.upsertEntries([record.data], toSoup: objectType.objectName))
     }
 
-    public func createEntry<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
+    public func locallyDeleteEntry(entry: objectType) {
+        var record: objectType = entry
         record.local = true
-        record.locallyCreated = true
-        syncEntry(entry: record, completion: completion)
+        record.locallyDeleted = true
+        _ = store.upsertEntries([record.data], toSoup: objectType.objectName)
+    }
+
+    public func createEntry(entry: objectType) -> Promise<objectType> {
+        _ = locallyCreateEntry(entry: entry)
+        return syncUpDown()
+            .then { _ -> Promise<objectType> in
+                if let record = self.record(forExternalId: entry.externalId) {
+                    return Promise.value(record)
+                } else {
+                    return Promise(error:StoreErrors.recordNotFound)
```

Comment:
Thanks to promises, the code gets cleaner: instead of dealing with an optional that the caller would have to guard against, the caller is guaranteed to get the object in the then block. In case of error, the catch block of the promise chain will be invoked.